### PR TITLE
Rendezvous after receiving "exit" notification

### PR DESCRIPTION
**Pull request by MaskRay**
@lhmouse

### CODE REVIEW

File: src/pipeline.cc [hunk 1]
```diff
@@ -361,8 +366,31 @@ bool Indexer_Parse(SemaManager *completion, WorkingFiles *wfiles,
   return true;
 }
 
+void Quit(SemaManager &manager) {
+  quit = true;
+  manager.Quit();
+
+  { std::lock_guard lock(index_request->mutex_); }
+  indexer_waiter->cv.notify_all();
```
Comment (hotpxl) [inline src/pipeline.cc]: It’s better to protect the notify with the mutex. If the thread is woken up for any reason, the signal would be lost

File: src/sema_manager.cc [hunk 2]
```diff
@@ -432,6 +436,8 @@ void *CompletionMain(void *manager_) {
   set_thread_name("comp");
   while (true) {
     std::unique_ptr<SemaManager::CompTask> task = manager->comp_tasks.Dequeue();
+    if (pipeline::quit)
```
Comment (hotpxl) [inline src/sema_manager.cc]: A full memory barrier is probably not necessary here?

Comment (MaskRay) [reply src/sema_manager.cc]: `ThreadedQueue::{Dequeue,PushBack}` use a mutex internally, the access of `pipeline::quit` is guaranteed to be synchronized.

Comment (lhmouse) [reply src/sema_manager.cc]: Yes a seq_cst barrier is an overkill. Consider the following:

```c++
if(pipeline::quit.load(std::memory_order_relaxed)
  ...
````
(or `memory_order_acquire` if it wasn't protected by a mutex).

File: src/pipeline.cc (hunk 1 above)
Comment (lhmouse) [reply src/pipeline.cc]: The code is correct. It is a bit obscure, however.

File: src/sema_manager.cc (hunk 2 above)
Comment (MaskRay) [reply src/sema_manager.cc]: It seems everything is protected by a mutex. I can change every load of `quit` to use `std::memory_order_relaxed`

**Pull request merged**